packages/hash/frontend/src/components/hooks/useUpdateAuthenticatedUser.ts

```diff
@@ -0,0 +1,117 @@
+import { useMutation } from "@apollo/client";
+import { useCallback, useState } from "react";
+
+import { types } from "@hashintel/hash-shared/types";
+import { extractBaseUri } from "@blockprotocol/type-system-web";
+import { GraphQLError } from "graphql";
+import {
+  UpdatePersistedEntityMutation,
+  UpdatePersistedEntityMutationVariables,
+} from "../../graphql/apiTypes.gen";
+import { mustGetEntity, Subgraph } from "../../lib/subgraph";
+import { AuthenticatedUser, constructAuthenticatedUser } from "../../lib/user";
+import { updatePersistedEntityMutation } from "../../graphql/queries/knowledge/entity.queries";
+import { useAuthenticatedUser } from "./useAuthenticatedUser";
+
+type UpdateAuthenticatedUserParams = {
+  shortname?: string;
+  preferredName?: string;
+};
+
+export const useUpdateAuthenticatedUser = () => {
+  const { authenticatedUser, refetch } = useAuthenticatedUser();
+
+  const [updatePersistedEntity] = useMutation<
+    UpdatePersistedEntityMutation,
+    UpdatePersistedEntityMutationVariables
+  >(updatePersistedEntityMutation, { errorPolicy: "all" });
+
+  const [loading, setLoading] = useState<boolean>(false);
+
+  const updateAuthenticatedUser = useCallback(
+    async (
+      params: UpdateAuthenticatedUserParams,
+    ): Promise<{
+      updatedAuthenticatedUser?: AuthenticatedUser;
+      errors?: readonly GraphQLError[] | undefined;
+    }> => {
+      if (!params.shortname && !params.preferredName) {
+        return { updatedAuthenticatedUser: authenticatedUser };
+      }
+
+      const [
+        {
+          data: { me: latestSubgraph },
+        },
+      ] = await refetch();
+
+      const entityId = latestSubgraph.roots[0]!;
```

Comment:
As an alternative to passing around the ID and such, you could do:
`getRootsAsEntities(subgraph).pop()`

Reply:
I think I wasn't sure, or had forgotten, that we had a method called `extractEntityRoot` 😁 

My suggestion for `getRootsAsEntities` was just that it would handle all of the mapping and extraction of the inner of the vertex.

`extractEntityRoot` existing seems better, thanks